File: go/wallet/wallet_sdk.go
package wallet

import (
	"aesecb"
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

type Api struct {
	no        *string
	accessKey *string
	secretKey *string
	token     *string
	baseUrl   string
	client    *http.Client
}

type result struct {
	Code    int         `json:"code"`
	Data    interface{} `json:"data"`
	Message string      `json:"message"`
}

func (result *result) isSuccessful() bool {
	return result.Code == 200
}

type token struct {
	Token          string    `json:"token"`
	Expires        int64     `json:"expires"`
	ExpireDateTime time.Time `json:"expire_date_time"`
}

type Platform struct {
	No           string      `json:"no,omitempty"`
	AccessKey    string      `json:"accessKey,omitempty"`
	SecretKey    string      `json:"secretKey,omitempty"`
	Name         string      `json:"name,omitempty"`
	Wallet       string      `json:"wallet,omitempty"`
	Status       int         `json:"status,omitempty"`
	TransferRate json.Number `json:"transferRate,omitempty"`
	UpdateTime   string      `json:"updateTime,omitempty"`
	CreateTime   string      `json:"createTime,omitempty"`
}

type Asset struct {
	Symbol string `json:"symbol,omitempty"`
	//@SerializedName("contract", alternate = ["contractAddress"])
	Contract     string      `json:"contract,contractAddress,omitempty"`
	Name         string      `json:"name,omitempty"`
	Logo         string      `json:"logo,omitempty"`
	Number       json.Number `json:"number,omitempty"`
	Total        json.Number `json:"total,omitempty"`
	FreezeNumber json.Number `json:"freezeNumber,omitempty"`
	Introduce    string      `json:"introduce,omitempty"`
	WhiteBook    string      `json:"whiteBook,omitempty"`
	Status       int         `json:"status,omitempty"`
	UpdateTime   string      `json:"updateTime,omitempty"`
	CreateTime   string      `json:"createTime,omitempty"`

	From   string `json:"from,omitempty"`
	To     string `json:"to,omitempty"`
	Hash   string `json:"hash,omitempty"`
	Remark string `json:"remark,omitempty"`
}

func CreateSDK(no string, accessKey string, secretKey string) *Api {
	api := Api{
		client: &http.Client{
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
			},
		},
		baseUrl:   "https://wallet.codbtoken.com/api",
		no:        &no,
		accessKey: &accessKey,
		secretKey: &secretKey,
	}
	return &api
}

func (api *Api) postJson(_url string, body interface{}, retry bool) []byte {
	data, ok := body.(string)
	var reader io.Reader
	if ok {
		if strings.HasSuffix(_url, "platform/token") {
			reader = strings.NewReader(data)
		} else {
			reader = strings.NewReader(bodyAesECBEncrypt(*api.secretKey, data))
		}
	} else {
		buffer, _ := json.Marshal(body)
		if strings.HasSuffix(_url, "platform/token") {
			reader = strings.NewReader(string(buffer))
		} else {
			reader = strings.NewReader(bodyAesECBEncrypt(*api.secretKey, string(buffer)))
		}
	}
	req, _ := http.NewRequest("POST", _url, reader)
	req.Header.Add("Content-Type", "application/json")
	if api.token != nil {
		req.Header.Add("token", *api.token)
	}
	response, err := api.client.Do(req)
	if err != nil {
		fmt.Printf("request faile: %v", err)
	} else {
		if response.StatusCode == 200 {
			buffer, _ := ioutil.ReadAll(response.Body)
			return buffer
		} else if response.StatusCode == 401 {
			result := api.getToken()
			if result != nil {
				token, ok := result.Data.(*token)
				if ok {
					api.token = &token.Token
					if retry {
						return api.postJson(_url, body, false)
					}
				}
			}
		}
	}
	return nil
}

func (api *Api) get(_url string, query map[string]string, retry bool) []byte {
	url_, err := url.Parse(_url)
	if url_ == nil {
		return nil
	}
	if query != nil {
		for key := range query {
			url_.Query().Add(key, query[key])
		}
	}
	req, _ := http.NewRequest("GET", url_.String(), nil)
	if api.token != nil {
		req.Header.Add("token", *api.token)
	}
	response, err := api.client.Do(req)
	if err != nil {
		fmt.Printf("request faile: %v", err)
	} else {
		if response.StatusCode == 200 {
			buffer, _ := ioutil.ReadAll(response.Body)
			return buffer
		} else if response.StatusCode == 401 {
			result := api.getToken()
			if result != nil {
				token, ok := result.Data.(*token)
				if ok {
					api.token = &token.Token
					if retry {
						return api.get(_url, query, false)
					}
				}
			}
		}
	}
	return nil
}

func (api *Api) url(path string) string {
	return fmt.Sprintf("%s%s", api.baseUrl, path)
}

func (api *Api) unmarshal(data []byte, typeStruct interface{}) bool {
	var result = &result{}
	_ = json.Unmarshal(data, result)
	encBuffer := result.Data.(string)
	if result.isSuccessful() && len(encBuffer) > 0 {
		decBuffer := bodyAesECBDecrypt(*api.secretKey, encBuffer)
		err := json.Unmarshal(decBuffer, typeStruct)
		if err == nil {
			return true
		}
	}
	return false
}

func (api *Api) getToken() *result {
	var body = map[string]string{
		"no":        *api.no,
		"accessKey": *api.accessKey,
	}
	data := api.postJson(api.url("/platform/token"), body, true)
	if data != nil {
		var result = &result{
			Data: &token{},
		}
		_ = json.Unmarshal(data, result)
		return result
	}
	return nil
}

// Publish 发布资产
// @param symbol 资产名（只能使用字母）
// @param name 资产别名（可以用中文0
// @param total 发布数量
// @param logo 资产logo
// @param introduce 资产介绍
// @param whiteBook 白皮书
// @return 调用远程接口是否成功
func (api *Api) Publish(symbol string, name string, total string, logo string, introduce string, whiteBook string) bool {
	data := api.postJson(api.url("/platform/publish"), map[string]string{
		"symbol":    symbol,
		"name":      name,
		"total":     total,
		"logo":      logo,
		"introduce": introduce,
		"whiteBook": whiteBook,
	}, true)
	if data != nil {
		var result = &result{}
		_ = json.Unmarshal(data, result)
		if result.isSuccessful() {
			return true
		}
	}
	return false
}

// Transfer 转帐
// @param contract 资产合约地址
// @param from 付款地址
// @param to 收款地址
// @param amount 转帐数量
// @param remark 备注
// @return 调用远程接口是否成功
func (api *Api) Transfer(contract string, from string, to string, amount string, remark string) bool {
	data := api.postJson(api.url("/platform/asset/transfer"), map[string]string{
		"contract": contract,
		"from":     from,
		"to":       to,
		"amount":   amount,
		"remark":   remark,
	}, true)
	if data != nil {
		var result = &result{}
		_ = json.Unmarshal(data, result)
		if result.isSuccessful() {
			return true
		}
	}
	return false
}

// RegisterUser 注册平台用户
// @param uid 该平台用户唯一标识，不能重复
// @return 该用户钱包地址
func (api *Api) RegisterUser(uid string) *string {
	data := api.postJson(api.url("/platform/registerUser"), map[string]string{"uid": uid}, true)
	if data != nil {
		var result = &result{}
		_ = json.Unmarshal(data, result)
		str, ok := result.Data.(string)
		if ok {
			buffer := bodyAesECBDecrypt(*api.secretKey, string(str))
			if buffer != nil {
				var address = string(buffer)
				return &address
			}
		}
	}
	return nil
}

// UpdateWalletAssetStatus 操作指定钱包资产数量
// @param wallet 指定的钱包地址
// @param contract 指定的合约地址
// @param action 0: 解除冻结, 1: 冻结
// @param amount 解除冻结或者冻结数量
// @return 调用远程接口是否成功
func (api *Api) UpdateWalletAssetStatus(wallet string, contract string, action int, amount string) bool {
	data := api.postJson(api.url("/platform/asset/updateUserAssetStatus"), map[string]interface{}{
		"contract": contract,
		"wallet":   wallet,
		"action":   action,
		"amount":   amount,
	}, true)
	if data != nil {
		var result = &result{}
		_ = json.Unmarshal(data, result)
		if result.isSuccessful() {
			return true
		}
	}
	return false
}

// GetPlatformInfo 获取平台信息
// @return 平台信息封装对象
func (api *Api) GetPlatformInfo() *Platform {
	data := api.get(api.url("/platform/info"), nil, true)
	if data != nil {
		var platform = &Platform{}
		if api.unmarshal(data, platform) {
			return platform
		}
	}
	return nil
}

// GetPlatformAssets 获取平台资产列表
// @return 平台资产列表
func (api *Api) GetPlatformAssets() []*Asset {
	data := api.get(api.url("/platform/asset/listPlatformAssets"), nil, true)
	if data != nil {
		println(string(data))
		var assets []*Asset
		if api.unmarshal(data, &assets) {
			return assets
		}
	}
	return nil
}

// GetWalletAssetLog 获取指定钱包资产变动记录
// @param address 要获取的钱包地址
// @param contract 合约地址, 不传则是平台下的所有资产变动
// @param page 页码，1开始
// @param limit 每页数量
// @return 资产变动列表
func (api *Api) GetWalletAssetLog(address string, contract string, page int, limit int) []*Asset {
	data := api.get(api.url("/platform/asset/userAssetLog"), map[string]string{
		"address":  address,
		"contract": contract,
		"page":     strconv.Itoa(page),
		"limit":    strconv.Itoa(limit),
	}, true)
	if data != nil {
		var assets []*Asset
		if api.unmarshal(data, assets) {
			return assets
		}
	}
	return nil
}

// GetWalletAssets 获取指定钱包持有资产列表
// @param address 指定的钱包地址
// @return 资产列表
func (api *Api) GetWalletAssets(address string) []*Asset {
	data := api.get(api.url("/platform/asset/listUserAssets"), map[string]string{"address": address}, true)
	if data != nil {
		var assets []*Asset
		if api.unmarshal(data, assets) {
			return assets
		}
	}
	return nil
}

// GetWalletAsset 获取指定钱包指定资产持有信息
// @param address 指定的钱包地址
// @param contract 指定的合约地址
// @return 资产信息
func (api *Api) GetWalletAsset(address string, contract string) *Asset {
	data := api.get(api.url("/platform/asset/userAsset"), map[string]string{"address": address, "contract": contract}, true)
	if data != nil {
		var asset = &Asset{}
		if api.unmarshal(data, asset) {
			return asset
		}
	}
	return nil
}

// GetAssetInfo 获取资产详情
// @param contract 指定的合约地址
// @return 资产详情
func (api *Api) GetAssetInfo(contract string) *Asset {
	data := api.get(api.url("/platform/asset/info"), map[string]string{"contract": contract}, true)
	if data != nil {
		var asset = &Asset{}
		if api.unmarshal(data, asset) {
			return asset
		}
	}
	return nil
}

func bodyAesECBDecrypt(key string, base64Text string) []byte {
	err := aesecb.SetAesKey(key)
	if err != nil {
		fmt.Printf("aes key初始化失败 %s\n", err)
	}
	sed, err := base64.StdEncoding.DecodeString(base64Text)
	if err != nil {
		fmt.Printf("base64 解码失败 %s\n", err)
	}
	sd, err := aesecb.AesECBDecrypt(sed)
	if err != nil {
		fmt.Printf("aes 解密失败 %s\n", err)
	}
	return sd
}

func bodyAesECBEncrypt(key string, date string) string {
	err := aesecb.SetAesKey(key)
	if err != nil {
		fmt.Printf("aes key初始化失败 %s\n", err)
	}
	se, err := aesecb.AesECBEncrypt([]byte(date))
	if err != nil {
		fmt.Printf("aes 加密失败 %s\n", err)
	}
	s := base64.StdEncoding.EncodeToString(se)
	return s
}

File: php/wallet_sdk.php
<?php

class ApiResponse
{
    private $code;
    private $data;
    private $message;

    private function __construct($code, $data, $message)
    {
        $this->code = $code;
        $this->data = $data;
        $this->message = $message;
        if ($message != null) {
            printf(" error: $message\r\n");
        } else {
            printf("\r\n");
        }
    }

    public static function create($code, $data, $message): ApiResponse
    {
        return new ApiResponse($code, $data, $message);
    }

    function code(): int
    {
        return $this->code;
    }

    function data()
    {
        return $this->data;
    }

    function getDecryptData(string $key)
    {
        $decrypt = openssl_decrypt(base64_decode($this->data, true), "aes-128-ecb", $key, OPENSSL_RAW_DATA);
        if ($decrypt) {
            $mixed = json_decode($decrypt);
            if ($mixed != null) {
                return $mixed;
            }
        }
        return $decrypt;
    }

    function isSuccessful(): bool
    {
        return $this->code == 200;
    }

    function error(): string
    {
        return $this->message;
    }
}

class WalletSDK
{
    protected $no;
    protected $accessKey;
    protected $secretKey;
    private $baseUrl = "https://wallet.codbtoken.com/api";
    private $token;

    private function __construct($no, $accessKey, $secretKey)
    {
        $this->no = $no;
        $this->accessKey = $accessKey;
        $this->secretKey = $secretKey;
    }

    public static function create($no, $accessKey, $secretKey): WalletSDK
    {
        return new WalletSDK($no, $accessKey, $secretKey);
    }

    private function sendRequest(string $url, $data, string $method, $contentType, int $timeout = 30, bool $proxy = false): ?ApiResponse
    {
        $ch = null;
        $ret = null;
        $header = array();
        $header[] = 'token: ' . $this->token;
        if ('POST' === strtoupper($method)) {
            $ch = curl_init($url);
            curl_setopt($ch, CURLOPT_POST, true);
            curl_setopt($ch, CURLOPT_HEADER, false);
            curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);
            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
            curl_setopt($ch, CURLOPT_FRESH_CONNECT, true);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_FORBID_REUSE, true);
            curl_setopt($ch, CURLOPT_TIMEOUT, $timeout);
            if ($contentType != null) {
                $header[] = 'Content-Type: ' . $contentType;
            }
            curl_setopt($ch, CURLOPT_HTTPHEADER, $header);
            if (is_string($data)) {
                curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
            } else {
                curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($data));
            }
        } else if ('GET' === strtoupper($method)) {
            if ($data != null) {
                if (is_string($data)) {
                    $real_url = $url . (strpos($url, '?') === false ? '?' : '') . $data;
                } else {
                    $real_url = $url . (strpos($url, '?') === false ? '?' : '') . http_build_query($data);
                }
            } else {
                $real_url = $url;
            }
            $ch = curl_init($real_url);
            curl_setopt($ch, CURLOPT_HEADER, false);
            curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);
            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
            curl_setopt($ch, CURLOPT_HTTPHEADER, $header);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_TIMEOUT, $timeout);
        } else {
            return ApiResponse::create(-1, null, "not support method: $method");
        }

        if ($proxy) {
            curl_setopt($ch, CURLOPT_PROXY, $proxy);
        }
        printf("$method: $url -> ");
        $ret = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        printf("http code: $httpCode");
        curl_close($ch);
        if ($ret && $httpCode == 200) {
//            printf("$ret\r\n");
            $obj = json_decode($ret);
            return ApiResponse::create($obj->code, $obj->data, $obj->message);
        } else {
            return ApiResponse::create($httpCode, null, "request fail");
        }
    }

    private function get(string $url, $data): ApiResponse
    {
        $response = $this->sendRequest($url, $data, "GET", null);
        if ($response->code() == 401) {
            $this->updateToken();
        } else {
            return $response;
        }
        return $this->sendRequest($url, $data, "GET", null);
    }

    private function postJson(string $url, $data, bool $encrypt = true): ApiResponse
    {
        if ($data != null) {
            if (!is_string($data)) {
                $data = json_encode($data);
                if (!$data) {
                    $data = null;
                } else if ($encrypt) {
                    $data = base64_encode(openssl_encrypt($data, "aes-128-ecb", $this->secretKey, OPENSSL_RAW_DATA));
                }
            }
        }
        $response = $this->sendRequest($url, $data, "POST", "application/json");
        if ($response->code() == 401) {
            $this->updateToken();
        } else {
            return $response;
        }
        return $this->sendRequest($url, $data, "POST", "application/json");
    }

    private function updateToken(): bool
    {
        $data = array("no" => $this->no, "accessKey" => $this->accessKey);
        $response = $this->postJson("$this->baseUrl/platform/token", $data, false);
        if ($response->isSuccessful()) {
            if (is_object($response->data())) {
                $this->token = $response->data()->token;
            }
        }
        return false;
    }

    /**
     * 发布资产
     * @param string $symbol 资产名（只能使用字母）
     * @param string $name 资产别名（可以用中文0
     * @param string $total 发布数量
     * @param string $logo 资产logo
     * @param string $introduce 资产介绍
     * @param string $whiteBook 白皮书
     * @return bool 调用远程接口是否成功
     */
    function publish(string $symbol, string $name, string $total, string $logo, string $introduce, string $whiteBook): bool
    {
        $data = array(
            "symbol" => $symbol,
            "name" => $name,
            "total" => $total,
            "logo" => $logo,
            "introduce" => $introduce,
            "whiteBook" => $whiteBook
        );
        $response = $this->postJson("$this->baseUrl/platform/publish", $data);
        return $response->isSuccessful();
    }

    /**
     * 转帐
     * @param string $contract 资产合约地址
     * @param string $from 付款地址
     * @param string $to 收款地址
     * @param string $amount 转帐数量
     * @param string $remark 备注
     * @param string $gasContract 平台收费使用的资产
     * @param string $gasFee 费用
     * @return bool 调用远程接口是否成功
     */
    function transfer(string $contract, string $from, string $to, string $amount, string $remark, string $gasContract, string $gasFee): bool
    {
        $data = array(
            "contract" => $contract,
            "from" => $from,
            "to" => $to,
            "amount" => $amount,
            "remark" => $remark
        );
        if ($gasContract && $gasFee) {
            $data["gas"] = array(
                "contract" => $gasContract,
                "free" => $gasFee
            );;
        }
        $response = $this->postJson("$this->baseUrl/platform/asset/transfer", $data);
        return $response->isSuccessful();
    }

    /**
     * 注册平台用户
     * @param string $uid 该平台用户唯一标识，不能重复
     * @return string|null 该用户钱包地址
     */
    function registerUser(string $uid): ?string
    {
        $response = $this->postJson("$this->baseUrl/platform/registerUser", array("uid" => $uid));
        if ($response->isSuccessful()) {
            $data = $response->getDecryptData($this->secretKey);
            return (string)$data;
        }
        return null;
    }

    /**
     * 操作指定钱包资产数量
     * @param string $wallet 指定的钱包地址
     * @param string $contract 指定的合约地址
     * @param string $action 0: 解除冻结, 1: 冻结
     * @param string $amount 解除冻结或者冻结数量
     * @return bool 调用远程接口是否成功
     */
    function updateWalletAssetStatus(string $wallet, string $contract, string $action, string $amount): bool
    {
        $data = array(
            "wallet" => $wallet,
            "contract" => $contract,
            "action" => $action,
            "amount" => $amount
        );
        $response = $this->postJson("$this->baseUrl/platform/asset/updateUserAssetStatus", $data);
        return $response->isSuccessful();
    }

    /**
     * 获取平台信息
     * @return Platform|null 平台信息封装对象
     */
    function getPlatformInfo(): ?Platform
    {
        $response = $this->get("$this->baseUrl/platform/info", null);
        if ($response->isSuccessful()) {
            $data = $response->getDecryptData($this->secretKey);
            if (is_object($data)) {
                return Platform::build($data);
            }
        }
        return null;
    }

    /**
     * 获取平台用户信息
     * @param string $uid 平台用户ID
     * @return object 平台用户信息
     */
    function getPlatformUserInfo(string $uid): ?object
    {
        $response = $this->get("$this->baseUrl/platform/platformUserInfo", array(
            "uid" => $uid,
        ));
        if ($response->isSuccessful()) {
            $data = $response->getDecryptData($this->secretKey);
            if (is_object($data)) {
                return (object) $data;
            }
        }
        return null;
    }

    /**
     * 获取平台资产列表
     * @return array 平台资产列表
     * if($asset instanceof Asset){
     *   printf($asset->symbol);
     * }
     */
    function getPlatformAssets(): array
    {
        $response = $this->get("$this->baseUrl/platform/asset/listPlatformAssets", null);
        $assets = array();
        if ($response->isSuccessful()) {
            $data = $response->getDecryptData($this->secretKey);
            if (is_array($data)) {
                foreach ($data as $value) {
                    $assets[] = Asset::build($value);
                }
            }
        }
        return $assets;
    }

    /**
     * 获取指定钱包资产变动记录
     * @param string $address 要获取的钱包地址
     * @param string $contract 合约地址, 不传则是平台下的所有资产变动
     * @param int $page 页码，1开始
     * @param int $limit 每页数量
     * @return array 资产变动列表
     */
    function getWalletAssetLog(string $address, string $contract, int $page, int $limit): array
    {
        $response = $this->get("$this->baseUrl/platform/asset/userAssetLog", array(
            "address" => $address,
            "contract" => $contract,
            "page" => $page,
            "limit" => $limit,
        ));
        $assets = array();
        if ($response->isSuccessful()) {
            $data = $response->getDecryptData($this->secretKey);
            if (is_array($data)) {
                foreach ($data as $value) {
                    $assets[] = Asset::build($value);
                }
            }
        }
        return $assets;
    }

    /**
     * 获取指定钱包持有资产列表
     * @param string $address 指定的钱包地址
     * @return array 资产列表
     */
    function getWalletAssets(string $address): array
    {
        $response = $this->get("$this->baseUrl/platform/asset/listUserAssets", array("address" => $address));
        $assets = array();
        if ($response->isSuccessful()) {
            $data = $response->getDecryptData($this->secretKey);
            if (is_array($data)) {
                foreach ($data as $value) {
                    $assets[] = Asset::build($value);
                }
            }
        }
        return $assets;
    }

    /**
     * 获取指定钱包指定资产持有信息
     * @param string $address 指定的钱包地址
     * @param string $contract 指定的合约地址
     * @return Asset|null 资产信息
     */
    function getWalletAsset(string $address, string $contract): ?Asset
    {
        $response = $this->get("$this->baseUrl/platform/asset/userAsset", array(
            "address" => $address,
            "contract" => $contract
        ));
        if ($response->isSuccessful()) {
            $data = $response->getDecryptData($this->secretKey);
            return Asset::build($data);
        }
        return null;
    }

    /**
     * 获取资产详情
     * @param string $contract 指定的合约地址
     * @return Asset|null 资产详情
     */
    function getAssetInfo(string $contract): ?Asset
    {
        $response = $this->get("$this->baseUrl/platform/asset/info", array("contract" => $contract));
        if ($response->isSuccessful()) {
            $data = $response->getDecryptData($this->secretKey);
            return Asset::build($data);
        }
        return null;
    }
}

class Platform
{
    var $no;
    var $accessKey;
    var $secretKey;
    var $name;
    var $wallet;
    var $status;
    var $transferRate;
    var $updateTime;
    var $createTime;

    public static function build($data): Platform
    {
        $platform = new Platform();
        $platform->no = $data->symbol;
        $platform->accessKey = $data->accessKey;
        $platform->secretKey = $data->secretKey;
        $platform->name = $data->name;
        $platform->wallet = $data->wallet;
        $platform->status = $data->status;
        $platform->transferRate = $data->transferRate;
        $platform->updateTime = $data->updateTime;
        $platform->createTime = $data->createTime;
        return $platform;
    }
}

class Asset
{
    var $symbol;
    var $contract;
    var $name;
    var $logo;
    var $number;
    var $total;
    var $freezeNumber; //冻结总数量
    var $useFreezeNumber; //质押数量
    var $introduce;
    var $whiteBook;
    var $status;
    var $updateTime;
    var $createTime;

    var $from;
    var $to;
    var $hash;
    var $remark;

    public static function build($data): Asset
    {
        $asset = new Asset();
        $asset->symbol = $data->symbol;
        $asset->contract = $data->contract;
        $asset->name = $data->name;
        $asset->logo = $data->logo;
        $asset->number = $data->number;
        $asset->total = $data->total;
        $asset->freezeNumber = $data->freezeNumber;
        $asset->useFreezeNumber = $data->useFreezeNumber;
        $asset->introduce = $data->introduce;
        $asset->whiteBook = $data->whiteBook;
        $asset->status = $data->status;
        $asset->updateTime = $data->updateTime;
        $asset->from = $data->from;
        $asset->to = $data->to;
        $asset->hash = $data->hash;
        $asset->remark = $data->remark;
        return $asset;
    }
}

//$api = WalletSDK::create("", "", "");
//print($api->getPlatformUserInfo("oapfJ5aJY_iWpl_mo3ZP0eH3-XSI")->wallet);
//$assets = $api->transfer("0x4cb89ac30f2342c89ca4025ba474083f4f205ca0", "0x3e754484ea14f521b02738c2aede37b527c4283d", "2", "3", "", "xx", "xxx");
//foreach ($assets as $asset) {
//    if ($asset instanceof Asset) {
//        printf("$asset->number \r\n");
//    }
//}